src/uct/base/uct_cm.c

```diff
@@ -82,28 +82,83 @@ ucs_status_t uct_cm_ep_pack_cb(uct_cm_base_ep_t *cep, void *arg,
+                                             uct_cm_base_ep_t *cep)
+{
+    if (params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR_CONNECT_CB) {
+        if (params->sockaddr_connect_cb.client != NULL) {
+            cep->client.connect_cb = params->sockaddr_connect_cb.client;
+        } else {
+            ucs_error("UCT_EP_PARAM_FIELD_SOCKADDR_CONNECT_CB is set but client's"
+                      "connect callback is NULL");
+            return UCS_ERR_INVALID_PARAM;
+        }
+    } else {
+        cep->client.connect_cb = NULL;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t uct_cm_ep_server_set_notify_cb(const uct_ep_params_t *params,
+                                            uct_cm_base_ep_t *cep)
+{
+    if (params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR_CONNECT_CB) {
+        if (params->sockaddr_connect_cb.client != NULL) {
+            cep->server.notify_cb = params->sockaddr_connect_cb.server;
+        } else {
+            ucs_error("UCT_EP_PARAM_FIELD_SOCKADDR_CONNECT_CB is set but server's"
+                      "notify callback is NULL");
+            return UCS_ERR_INVALID_PARAM;
+        }
+    } else {
+        cep->server.notify_cb = NULL;
+    }
+
+    return UCS_OK;
+}
+
 void uct_cm_ep_client_connect_cb(uct_cm_base_ep_t *cep,
                                  uct_cm_remote_data_t *remote_data,
                                  ucs_status_t status)
 {
     uct_cm_ep_client_connect_args_t connect_args;
 
-    connect_args.field_mask  = UCT_CM_EP_CLIENT_CONNECT_ARGS_FIELD_REMOTE_DATA |
-                               UCT_CM_EP_CLIENT_CONNECT_ARGS_FIELD_STATUS;
-    connect_args.remote_data = remote_data;
-    connect_args.status      = status;
+    if (cep->client.connect_cb != NULL) {
+        connect_args.field_mask  = UCT_CM_EP_CLIENT_CONNECT_ARGS_FIELD_REMOTE_DATA |
+                                   UCT_CM_EP_CLIENT_CONNECT_ARGS_FIELD_STATUS;
+        connect_args.remote_data = remote_data;
+        connect_args.status      = status;
 
-    cep->client.connect_cb(&cep->super.super, cep->user_data, &connect_args);
+        cep->client.connect_cb(&cep->super.super, cep->user_data, &connect_args);
+    }
 }
 
 void uct_cm_ep_server_conn_notify_cb(uct_cm_base_ep_t *cep, ucs_status_t status)
 {
     uct_cm_ep_server_conn_notify_args_t notify_args;
 
-    notify_args.field_mask = UCT_CM_EP_SERVER_CONN_NOTIFY_ARGS_FIELD_STATUS;
-    notify_args.status     = status;
+    if (cep->server.notify_cb != NULL) {
+        notify_args.field_mask = UCT_CM_EP_SERVER_CONN_NOTIFY_ARGS_FIELD_STATUS;
+        notify_args.status     = status;
+
+        cep->server.notify_cb(&cep->super.super, cep->user_data, &notify_args);
+    }
+}
+
+void uct_cm_set_common_data(uct_cm_base_ep_t *ep, const uct_ep_params_t *params)
+{
+    ep->priv_pack_cb  = (params->field_mask &
+                         UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                        params->sockaddr_pack_cb : NULL;
```

Comment:
i will add a check for priv_pack_cb (in another pr?)
disconnect_cb isn't checked since disconnect isn't implemented yet - will add it then
and user_data can be used even if it's null...

Reply:
ok